### PR TITLE
MP-26 prototype script export

### PR DESCRIPTION
**Pull request by JanionNevill**
In order to not make one enormous monolithic script, I have broken down the script into separate files, as they would be in a project. When you download the code, you are given a zip file containing the main script and an `R` folder containing the required functions. I opted to make it general enough that a different data file could be loaded into the script and it would still work regardless of whether it is long/wide or binary/continuous.

When you download the code, unzip it, then open the folder as a new project in R Studio. At the top of the main file, you need to load in a data file, then running the script will create a forest plot.

The export from each feature which is given this ability, will include the full required set of library functions by default. The user is also given options to download the full codebase , or just the script. This allows users to download the full codebase once, then just the individual scripts later, without creating duplicate files.

As originally scoped, this is a prototype, so any feedback on both code quality and user experience is warmly welcomed.

### CODE REVIEW

**Comment by JanionNevill** [comment]
I've pared it back a bit so that the data is included in the exported script, so it's fully runnable straight off the download. I've also added script downloads to the rest of the frequentist page.

Do you think that a second reviewer would be helpful on this?

**Review by WillRobinn — APPROVED** [review]
Frequentist works really well. The inclusion of the data in the script makes it super easy to run through the code.

A second reviewer would be good since this is a sizable update and is a new concept across all of the apps.

**Comment by JanionNevill** [comment]
@RyanJField I've requested your review as this is a pretty complex change which could use another set of eyes.

**Comment by JanionNevill** [comment]
Hi @RyanJField, do you have an estimate for when you might get to this?

File: server.R [hunk 1]
```diff
@@ -36,8 +36,12 @@ function(input, output, session) {
     )
     return(raw_data[rows, ])
   })
+  filtered_data <- shinymeta::metaReactive({
+    shinymeta::..(filtered_data_intermediate())
+  })
   
-  ContBin <- reactive({           # automatically detect if continuous or binary
+  ContBin_intermediate <- reactive({
```
Comment (RyanJField) [inline server.R:43]: Mixing naming conventions

File: R/script_generation.R [hunk 2]
```diff
@@ -0,0 +1,129 @@
+
+#' Gather and format code to be exported.
+#'
+#' @param meta_action The return from a call to `shinymeta::metaAction()`.
+#' @param expanded_chain The result from a call to `shinymeta::expandChain()`.
+#'
+#' @return Formatted code as text.
+.BuildMetaPairwiseExportCode <- function(meta_action = NULL, expanded_chain = NULL) {
```
Comment (RyanJField) [inline R/script_generation.R:8]: Are you using the leading . to differentiate functions?

**Review by RyanJField — COMMENTED** [review]
Couple of minor points.

There are some missing Roxgyen comments on some functions but I believe these were just moved.

The final exported script has a lot of libraries that are not needed to run the script, is there an easy way to remove them, or is it simply because they are captured by MetaActions? One way around this may be to use require and explicit scoping on the libraries not needed for export?

Finally we may need to add instructions to set the working directory to make sure the R directory can be correctly sourced (as the exported script is not in a project).

File: R/script_generation.R (hunk 2 above)
Comment (JanionNevill) [reply R/script_generation.R:8]: This is the standard for "private" functions, ie. those which are not to be called from outside this file.

File: server.R (hunk 1 above)
Comment (JanionNevill) [reply server.R:43]: 💀

**Comment by JanionNevill** [comment]
> The final exported script has a lot of libraries that are not needed to run the script, is there an easy way to remove them, or is it simply because they are captured by MetaActions? One way around this may be to use require and explicit scoping on the libraries not needed for export?

This is simply because they are captured by the MetaAction, so it's not trivial to capture only the relevant libraries. I have pulled out the shiny libraries though, as they are certainly not needed for the script. This does tidy it up a bit.

**Comment by JanionNevill** [comment]
> Finally we may need to add instructions to set the working directory to make sure the R directory can be correctly sourced (as the exported script is not in a project).

I've added a *README.md* file into the downloaded zip file so that users can run the script. Does that sufficiently address this?

**Comment by JanionNevill** [comment]
> There are some missing Roxgyen comments on some functions but I believe these were just moved.

Yeah, I checked those and found that only existing (moved) functions have missing Roxygen. Can I get a pass on these please?

**Comment by RyanJField** [comment]
> > Finally we may need to add instructions to set the working directory to make sure the R directory can be correctly sourced (as the exported script is not in a project).
> 
> I've added a _README.md_ file into the downloaded zip file so that users can run the script. Does that sufficiently address this?

Yes but one more problem, if were including run instructions it should include installing packages:
```
install.packages(
    "dplyr",
    "forcats",
    "ggplot2",
    "glue",
    "markdown",
    "metafor",
    "MetaStan",
    "purrr",
    "rintrojs",
    "rio",
    "rmarkdown",
    "rstan",
    "tidybayes",
    "tidyr",
    "tidyverse"
)
```

or something like: 
```
packages <- c(
    "dplyr",
    "forcats",
    "ggplot2",
    "glue",
    "markdown",
    "metafor",
    "MetaStan",
    "purrr",
    "rintrojs",
    "rio",
    "rmarkdown",
    "rstan",
    "tidybayes",
    "tidyr",
    "tidyverse"
)

# Install packages not yet installed
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
```

**Review by RyanJField — CHANGES_REQUESTED** [review]
See comment regarding libraries.

**Comment by JanionNevill** [comment]
I've changed the loading of packages to include installation.

**Review by RyanJField — APPROVED** [review]

**Pull request merged**